### PR TITLE
Extra functionality of Tensor class

### PR DESCRIPTION
**Pull request by davydden**
the two commits are currently part of https://github.com/dealii/dealii/pull/621/commits
Wolfang wanted at least one of them to be in separate PR, thus I create it.

### CODE REVIEW

File: include/deal.II/base/tensor.h [hunk 1]
```diff
@@ -761,10 +776,10 @@ std::ostream &operator << (std::ostream &out, const Tensor<rank_,1> &p)
  * @relates Tensor
  * @author Guido Kanschat, 2000
  */
-template <int dim, typename Number>
+template <int dim, typename Number, typename OtherNumber>
 inline
 Number contract (const Tensor<1,dim,Number> &src1,
-                 const Tensor<1,dim,Number> &src2)
+                 const Tensor<1,dim,OtherNumber> &src2)
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: The return type of this should now be `typename ProductType<Number,OtherNumber>::type`. Same in the `operator*`.

File: include/deal.II/base/symmetric_tensor.h [hunk 2]
```diff
@@ -628,6 +628,13 @@ class SymmetricTensor
   SymmetricTensor   operator - () const;
 
   /**
+   * Simple addition of a multiple of a tensor, i.e. *this += a*T,
+   * where the scalar is of the same base type and
+   * T is a symmetric tensor of double.
+   */
+  void add(const Number &a, const SymmetricTensor<rank,dim,double> &p);
```
Comment (bangerth) [inline include/deal.II/base/symmetric_tensor.h]: Why only add `SymmetricTensor<rank,dim,double>` and not `SymmetricTensor<rank,dim,Number>`?

**Comment by bangerth** [comment]
I don't recall the conversation about this. Where do you need the `add` functions?

**Comment by davydden** [comment]
There was no conversation about the `add` function, but I use it heavily in the PR cited above, see https://github.com/davydden/dealii/commit/2e30cc3ce22af786c6fea72ac32c4c3cd0414f1a

File: include/deal.II/base/symmetric_tensor.h (hunk 2 above)
Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: that was used to add to the resulting tensor a number multiplied by the tensor object coming from the shape functions, thus `double` only

**Comment by davydden** [comment]
I think `add()` shall be quicker than something like
```
derivatives[q_point] += value *
typename ProductType<Number,dealii::Tensor<order,spacedim> >::type(*shape_derivative_ptr++);
```
as it shall not require a temporary object.

**Comment by bangerth** [comment]
This is for small objects with a handful of elements. The compiler will optimize all of this out -- let's not worry about these micro-optimizations unless we have evidence that they are necessary.

So I would vote to just leave the existing code of the form
```
derivatives[q_point] += value *
typename ProductType<Number,dealii::Tensor<order,spacedim> >::type(*shape_derivative_ptr++);
```
in place. There is a cost to adding member functions (as this discussion shows) and it doesn't seem worth it at this point.

**Comment by davydden** [comment]
ok

**Comment by davydden** [comment]
will check that it compiles...

**Comment by davydden** [comment]
builds fine.

**Comment by bangerth** [comment]
Looks good.

**Comment by davydden** [comment]
in reference to #2033

**Pull request merged**